servoshell: Do not focus and raise new auxiliary WebDriver-created `WebView`s

yezhizhen opened the pull request:
For Desktop port of `request_open_auxiliary_webview`, stay on the original WebView if the request originates WebDriver. 

This is to make sure `webdriver_server::handle_new_window` does not focus the new window, according to spec. See https://github.com/servo/servo/blob/c7eba2dbbac24e8571d12ce686173a11b287c5ca/tests/wpt/tests/webdriver/tests/classic/new_window/new_window.py#L31-L37 

**To clarify**: this won't change the behaviour when user interacts, but only affects WebDriver [New Window](https://w3c.github.io/webdriver/#new-window).

Testing: `./mach test-wpt -r --log-raw "D:/servo log/all.txt" ./tests/wpt/tests/webdriver/tests/classic --product servodriver` based on 96b097303783ff12d4a1a7986260ad3c19a31837

yezhizhen committed:
Update `dom::windowproxy` comment

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>

yezhizhen committed:
Fix webdriver `NewWindow` behaviour

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>

yezhizhen commented:
cc @jdm @longvatrong111 @PotatoCP @xiaochengh 
I ran the test for 11 times, and only extract those that are stable in all 11 runs to avoid including intermittent results.

yezhizhen commented on tests/wpt/meta/webdriver/tests/classic/element_send_keys/form_controls.py.ini:
This one is due to a previous PR which didn't update test results. See https://github.com/servo/servo/pull/37224#pullrequestreview-2903871157

longvatrong111 commented on components/script/dom/windowproxy.rs:
```diff
@@ -466,26 +466,27 @@ impl WindowProxy {
     ) -> Fallible<Option<DomRoot<WindowProxy>>> {
-        // Step 4.
+        // Step 5
         let non_empty_target = match target.as_ref() {
```
Can you add the description for each step here and below too?
It may be a bit verbose  but the step number alone is not very useful, and the spec may change so it would give incorrect information.

mrobinson commented on components/constellation/constellation.rs:
```diff
@@ -4702,25 +4702,39 @@ where
-            WebDriverCommandMsg::NewWebView(webview_id, sender, load_sender) => {
-                let (chan, port) = match ipc::channel() {
+            WebDriverCommandMsg::NewWebView(
+                src_webview_id,
```
```suggestion
                originating_webview_id,
```

mrobinson commented on ports/servoshell/desktop/app_state.rs:
```diff
@@ -477,9 +478,6 @@ impl WebViewDelegate for RunningAppState {
             .build();
 
         webview.notify_theme_change(self.inner().window.theme());
-        webview.focus();
-        webview.raise_to_top(true);
```
Why are you removing the focus step here?

mrobinson replied:
> Unify the behavior, as said in PR message. Also required by webdriver.

I understand that you have made the behavior the same as the embedded version of servoshell, but you also need to justify in the PR description why this change is okay to make for the desktop port. Was it unnecessary before?

mrobinson replied:
> For the normal user interaction to open new tab, the behaviour won't change, as it is handled in another part.

Triggering `window.open()` is also normal user behavior here, I would say.

mrobinson commented on ports/servoshell/desktop/app_state.rs:
```diff
@@ -467,6 +467,7 @@ impl WebViewDelegate for RunningAppState {
         );
     }
 
+    // We open the auxiliary webview without focusing
```
I don't think this comment is necessary.

mrobinson commented on ports/servoshell/egl/app_state.rs:
```diff
@@ -237,6 +237,7 @@ impl WebViewDelegate for RunningAppState {
         }
     }
 
+    // We open the auxiliary webview without focusing
```
Please remove this comment. A comment should explain why something is important to do, not just describe what it does.

yezhizhen replied:
How adding this below:
"please don't change this behaviour because it is required by webdriver"

mrobinson replied:
Why is the behavior required by WebDriver? Is it in the specification?

yezhizhen replied:
Yeah. Opening new window must be done without invoking focusing steps. You can see the test in PR message.

mrobinson reviewed (changes requested):
Please rename the PR to describe the changes, not just their results ("make tests pass"). For instance, you might say something like, "Do not automatically focus new WebViews in servoshell." You will also need to describe why this isn't necessary for normal (non-WebDriver) interaction. Thanks!

yezhizhen commented:
> you might say something like, "Do not automatically focus new WebViews in servoshell." You will also need to describe why this isn't necessary for normal (non-WebDriver) interaction. Thanks!

I think this is browser specific. If you want, you can also make normal interaction not auto-focusing. But I think all browsers I've seen would auto focus the new tab.

mrobinson commented:
> I think this is browser specific. If you want, you can also make normal interaction not auto-focusing. But I think all browsers I've seen would auto focus the new tab. That's also what Servo is doing right now

So I'm a bit confused. Is it that we are also focusing when the tab becomes active and the initial focus during WebView creation is redundant?

yezhizhen commented on tests/wpt/meta/webdriver/tests/classic/element_send_keys/form_controls.py.ini:
This one is due to a previous PR which didn't update test results. See https://github.com/servo/servo/pull/37224#pullrequestreview-2903871157

mrobinson commented:
> The expected behaviour is in following video. I am not sure if I understand your question.

`raise_to_top()` and `focus()` are required to show new WebViews. You are removing the code from the creation of new WebViews. Where are they called? Are we guaranteed that they will always be called when using servoshell interactively?

yezhizhen commented:
> `raise_to_top()` and `focus()` are required to show new WebViews. You are removing the code from the creation of new WebViews. Where are they called? Are we guaranteed that they will always be called when using servoshell interactively?

Yes, I am very sure. I checked this part of code for a few days. 
When you interact as a human, `new_toplevel_webview` is called instead of `request_open_auxiliary_webview`, which still has these steps.

yezhizhen commented:
> **To clarify**: this won't change the behaviour when user opens a new tab; only changes it for `window.open` and WebDriver Open New Window. It only affects [window.open()](https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-open) and WebDriver [New Window](https://w3c.github.io/webdriver/#new-window).

@mrobinson I am not sure the new title you change to is suitable. It also affects `window.open()`. Maybe i will find a new one later.

yezhizhen commented:
> > Yes, I am very sure. I checked this part of code for a few days. When you interact as a human, `new_toplevel_webview` is called instead of `request_open_auxiliary_webview`, which still has these steps.
> 
> `new_toplevel_webview` is used when creating new `WebView`s, but `request_open_auxiliary_webview` is used for `window.open()`. Both are used when using servoshell interactively.

Yes. That's what I said in clarify in PR message.

mrobinson commented:
I think this is an example of an issue that makes it more obvious why we want to implement WebDriver on top of servoshell. For `window.open()` we want to activate the tab of new `WebView`s to match the behavior or browsers, but for WebDriver we do not want to do this. We cannot fix the WebDriver issue without breaking `window.open()` and it doesn't seem right to add any extra arguments to the embedder message as, in general, embedders shouldn't need to care about our internal WebDriver implementation.

yezhizhen committed:
Update Webdriver wpt-test

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>

mrobinson reviewed (changes requested):
Okay. This looks pretty good, with one change.

jdm commented:
> I think this is an example of an issue that makes it more obvious why we want to implement WebDriver on top of servoshell. For `window.open()` we want to activate the tab of new `WebView`s to match the behavior or browsers, but for WebDriver we do not want to do this. We cannot fix the WebDriver issue without breaking `window.open()` and it doesn't seem right to add any extra arguments to the embedder message as, in general, embedders shouldn't need to care about our internal WebDriver implementation.

This is actually a case where making this behaviour configurable to embedders makes sense to me. It's common in browsers to have an option to focus newly opened tabs/windows. We could add a preference that controls that behaviour, but we could also delegate that decision to the embedder, which is my preferred option.

mrobinson commented:
> Maybe wait a bit and check what others think? @mrobinson

I think we should preserve the behavior of servoshell, which was a deliberate choice, while also trying to fix the issue with WebDriver. It should be very straightforward here. If we want to modify the behavior of interactive servoshell, it should come with a justification for why the behavior is an improvement for users or is made due to some technical limitation.

mrobinson commented:
Here is a small diff which I believe preserves the existing servoshell behavior, but also modifies the behavior for WebDriver:

```diff
diff --git a/ports/servoshell/desktop/app_state.rs b/ports/servoshell/desktop/app_state.rs
index 464e344424a..5318af82917 100644
--- a/ports/servoshell/desktop/app_state.rs
+++ b/ports/servoshell/desktop/app_state.rs
@@ -12,7 +12,7 @@ use image::{DynamicImage, ImageFormat};
 use keyboard_types::{Key, KeyboardEvent, Modifiers, ShortcutMatcher};
 use log::{error, info};
 use servo::base::id::WebViewId;
-use servo::config::pref;
+use servo::config::{opts, pref};
 use servo::ipc_channel::ipc::IpcSender;
 use servo::webrender_api::ScrollLocation;
 use servo::webrender_api::units::{DeviceIntPoint, DeviceIntRect, DeviceIntSize};
@@ -477,8 +477,14 @@ impl WebViewDelegate for RunningAppState {
             .build();
 
         webview.notify_theme_change(self.inner().window.theme());
-        webview.focus();
-        webview.raise_to_top(true);
+
+        // When WebDriver is enabled, do not focus and raise the WebView to the top,
+        // as that is what the specification expects. Otherwise, we would like `window.open()`
+        // to create a new foreground tab
+        if opts::get().webdriver_port.is_some() {
+            webview.focus();
+            webview.raise_to_top(true);
+        }
 
         self.add(webview.clone());
         Some(webview)
```

yezhizhen committed:
Patch: Only change `window.open` behaviour for webdriver

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>

yezhizhen commented:
> I think we should preserve the behavior of servoshell, which was a deliberate choice, while also trying to fix the issue with WebDriver. It should be very straightforward here. If we want to modify the behavior of interactive servoshell, it should come with a justification for why the behavior is an improvement for users or is made due to some technical limitation.

Thanks! This sounds reasonable to me. Just added the logic, rebased on adb5df3354a7790e675faa8fbedf83c860d169bd, and updated test result.

yezhizhen committed:
Rebase and update wpt-test for `shadow_root`

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>

yezhizhen committed:
Further comments

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>

mrobinson reviewed (approved):
Looks good! Just a couple minor comments and a question.

yezhizhen committed:
Apply Martin's suggestion

Signed-off-by: Euclid Ye <yezhizhenjiakang@gmail.com>
